### PR TITLE
Add ROADMAP.md

### PR DESCRIPTION
**Pull request by icecrime**
Title says it all: this is a long overdue document describing what's next for the Docker Engine, as a hopefully faithful transcript of @shykes direction.

Ping @docker/core-maintainers as I think this is important!

### CODE REVIEW

**Comment by estesp** [comment]
:+1:

**Comment by tianon** [comment]
Love it :heart_eyes:

Seeing all this spelled out in writing is fantastic! :+1:

**Comment by calavera** [comment]
:dancer: :raised_hands: :heart: 

Everything sounds fantastic. LGTM

File: ROADMAP.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+### 1.2.2 Builder
+
+The Builder (i.e., the ability to build an image from a Dockerfile) is already nicely decoupled,
+but would benefit from being entirely separated from the Engine, and rely on the standard Engine
+API for its operations.
+
+### 1.2.3 Distribution
+
+Distribution already has a [dedicated repository](https://github.com/docker/distribution) which
```
Comment (stevvooe) [inline ROADMAP.md]: Let's add the following:

```
There are two short term goals related to image distribution. The first is stabilize and simplify the push/pull code. Following that is the conversion to the more secure Registry V2 protocol.
```

I'll work on making this section better when we get this road map clarified a little bit in the future.

Comment (icecrime) [inline ROADMAP.md]: Done (didn't invalidate the line though, so the comment just lingers here).

**Comment by stevvooe** [comment]
LGTM

**Comment by thaJeztah** [comment]
Thanks so much, @icecrime!

LGTM

This is an important document, I think we should merge this and touch up language in a later stage.

@moxiegirl hope you can agree on that?

File: ROADMAP.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+the near future are:
+
+- Trusted distribution of images: the effort is driven by the [distribution](https://github.com/docker/distribution)
+group but will have significant impact on the Engine
+- [User namespaces](https://github.com/docker/docker/pull/12648)
+- [Seccomp support](https://github.com/docker/libcontainer/pull/613)
+
+## 1.2 Plumbing project
+
+We define an plumbing tool as a standalone piece of software usable and meaningful on its own. In
```
Comment (duglin) [inline ROADMAP.md]: nit,  s/an plumbing/a plumbing/

File: ROADMAP.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+This plan will deprecate the existing [`execdriver`](https://github.com/docker/docker/tree/master/daemon/execdriver)
+as different runtime backends will be implemented as separated binaries instead of being compiled
+into the Engine.
+
+### 1.2.2 Builder
+
+The Builder (i.e., the ability to build an image from a Dockerfile) is already nicely decoupled,
+but would benefit from being entirely separated from the Engine, and rely on the standard Engine
+API for its operations.
```
Comment (duglin) [inline ROADMAP.md]: Maybe point to Josh's CLI-side builder PR/branch?  I don't have the url right now

File: ROADMAP.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+### 1.2.3 Distribution
```
Comment (duglin) [inline ROADMAP.md]: Seems like Distribution should have its own roadmaplist-of-work-items - and if so we can remove this from here.

**Comment by moxiegirl** [comment]
@thaJeztah Roadmap is a planning document.  The language is totally the purview of the engineering team, the maintainers, and the project management team. IOW, docs shouldn't review this or tweak this like something that goes onto docs.docker.com.  (Though happy to do so if they ask of course!)

File: ROADMAP.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+deliberately keep the extensions API the simplest possible, expanding as we discover valid use
+cases that cannot be implemented.
+
+At the time of writing:
+
+- Plugin support is merged as an experimental feature: real world use cases and user feedback will
+help us refine the UX to make the feature more user friendly.
+- There are no immediate plans to expand on the number of pluggable subsystems.
+- Golang 1.5 might add language support for [plugins](https://docs.google.com/document/d/1nr-TQHw_er6GOQRsF6T43GGhFDelrAP0NqSS_00RgZQ)
+which we consider supporting as an alternative to JSON/HTTP.
```
Comment (duglin) [inline ROADMAP.md]: Might want to say something about CLI plugins too - I believe I heard that the CLI will probably use the golang stuff too.

File: ROADMAP.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+## 1.5 Better API implementation
+
+The current Engine API is insufficiently typed, versioned, and ultimately hard to maintain. We
+also suffer from the lack of a common implementation with [Swarm](https://github.com/docker/swarm).
+
+## 1.6 Checkpoint/restore
+
+Support for checkpoint/restore was [merged](https://github.com/docker/libcontainer/pull/479) in
+[libcontainer](https://github.com/docker/libcontainer): we intend to take advantage of it in the
```
Comment (duglin) [inline ROADMAP.md]: s/libcontainer/runc/   maybe?

File: ROADMAP.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+- Long term impact of syntax changes is a sensitive matter that require an amount of attention
+the volume of Engine codebase and activity today doesn't allow us to provide
+- Allowing the Builder to be implemented as a separate utility consuming the Engine's API will
+open the door for many possibilities, such as offering alternate syntaxes or DSL for existing
+languages without cluttering the Engine's codebase
+- A standalone Builder will also offer the opportunity for a better dedicated group of maintainers
+to own the Dockerfile syntax and decide collectively on the direction to give it
+- Our experience with official images tend to show that no new instruction or syntax expansion is
+*strictly* necessary for the majority of use cases, and although we are aware many things are still
+lacking for many, we cannot make it a priority yet for the above reasons.
```
Comment (duglin) [inline ROADMAP.md]: I'd like to see the new CLI-side builder effort use that switch-over as an opportunity to clean-up the Dockerfile syntax. For example, get rid of either COPY or ADD and have more consistent parsing (e.g. use a space/quote aware parser for all commands).

**Comment by thaJeztah** [comment]
Thanks @moxiegirl, that was my general idea as well; so @icecrime feel free to merge after the "content" is agreed on (I see some notes have been added above)

File: ROADMAP.md (hunk 3 above)
Comment (icecrime) [inline ROADMAP.md]: Makes sense: I'll point to it in the associated issue.

File: ROADMAP.md (hunk 4 above)
Comment (icecrime) [inline ROADMAP.md]: I think [it does](https://github.com/docker/distribution/blob/master/ROADMAP.md), but some of them still have impact in the Engine.

File: ROADMAP.md (hunk 5 above)
Comment (icecrime) [inline ROADMAP.md]: I personally heard nothing about CLI plugins :-( The only thing I know for sure is that Solomon has ongoing work on a "smart" client, but you might know more than I do.

**Comment by cpuguy83** [comment]
LG(reat!)TM

File: ROADMAP.md (hunk 5 above)
Comment (thaJeztah) [inline ROADMAP.md]: Think *couch* he's referring to https://github.com/docker/docker/pull/13140 ;-)

Comment (duglin) [inline ROADMAP.md]: As much as I'd like to be referring to that  :-)  I wasn't.  I heard rumors of using go's v1.5 plugin stuff.

Comment (thaJeztah) [inline ROADMAP.md]: Ah, lol, thought you were pitching your PR (which would also be fine :-))

**Pull request merged**